File: app/src/main/kotlin/tech/mmarca/openvitals/domain/insights/PeriodComparison.kt
package tech.mmarca.openvitals.domain.insights

import kotlin.math.abs

private const val ComparisonTolerance = 0.0001

enum class PeriodComparisonDirection {
    UP,
    DOWN,
    SAME,
}

data class PeriodComparison(
    val currentValue: Double,
    val previousValue: Double,
) {
    val change: Double get() = currentValue - previousValue
    val absoluteChange: Double get() = abs(change)
    val percentChange: Double?
        get() = previousValue.takeIf { abs(it) > ComparisonTolerance }
            ?.let { change / it * 100.0 }
    val direction: PeriodComparisonDirection
        get() = when {
            change > ComparisonTolerance -> PeriodComparisonDirection.UP
            change < -ComparisonTolerance -> PeriodComparisonDirection.DOWN
            else -> PeriodComparisonDirection.SAME
        }
}

fun periodComparison(
    currentValue: Double,
    previousValue: Double,
): PeriodComparison =
    PeriodComparison(
        currentValue = currentValue,
        previousValue = previousValue,
    )

File: lib/domain/insights/period_comparison.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'period_comparison.freezed.dart';

const double _comparisonTolerance = 0.0001;

enum PeriodComparisonDirection {
  up,
  down,
  same,
}

@freezed
abstract class PeriodComparison with _$PeriodComparison {
  const PeriodComparison._();

  const factory PeriodComparison({
    required double currentValue,
    required double previousValue,
  }) = _PeriodComparison;

  double get change => currentValue - previousValue;

  double get absoluteChange => change.abs();

  double? get percentChange => previousValue.abs() > _comparisonTolerance
      ? change / previousValue * 100.0
      : null;

  PeriodComparisonDirection get direction {
    if (change > _comparisonTolerance) return PeriodComparisonDirection.up;
    if (change < -_comparisonTolerance) return PeriodComparisonDirection.down;
    return PeriodComparisonDirection.same;
  }
}

PeriodComparison periodComparison(
  double currentValue,
  double previousValue,
) =>
    PeriodComparison(
      currentValue: currentValue,
      previousValue: previousValue,
    );
